keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,456 @@
+improve the pod to node isolation (phase 1 and 2) and pod to pod isolation
+(phase 3) we currently have.
+
+This mitigates all the vulnerabilities listed in the motivation section.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I want run some pods with privileged capabilities because
+the applications in the pods require it (e.g. `CAP_SYS_ADMIN` to mount a FUSE
+filesystem or `CAP_NET_ADMIN` to setup a VPN) but I don't want this extra
+capability to grant me any extra privileges on the host or other pods.
+
+#### Story 2
+
+As a cluster admin, I want to allow some pods to run in the host user namespace
+if they need a feature only available in such user namespace, such as loading a
+kernel module with `CAP_SYS_MODULE`.
+
+#### Story 3
+
+As a cluster admin, I want to allow users to run their container as root
+without that process having root privileged on the host, so I can mitigate the
+impact of a compromised container.
+
+#### Story 4
+
+As a cluster admin, I want to allow users to choose the user/group they use
+inside the container without it being the user/group they use on the node, so I
+can mitigate the impact of a container breakout vulnerability (e.g to read
+host files).
+
+#### Story 5
+
+As a cluster admin, I want to use different host UIDs/GIDs for pods running in
+the same node (whenever kernel/kube features allow it), so I can mitigate the
+impact a compromised pod can have on other pods and the node itself.
+
+### Notes/Constraints/Caveats
+
+#### UID space
+
+#### Volume support
+
+### Risks and Mitigations
+
+## Design Details
+
+Note: Names are preliminary yet, I'm using field names to simplify explanations.
+
+### Pod.spec changes
+
+The following changes will be done to the pod.spec:
+
+- `pod.spec.useHostUsers`: bool.
+If true or not present, uses the host user namespace (as today)
+If false, a new userns is created for the pod.
+This field will be used for phase 1, 2 and 3.
+
+- `pod.spec.securityContext.userns.pod2podIsolation`: Enum
+If enabled, we will make the userns mappings be non-overlapping as much as possible.
+This field will be used in phase 3.
+
+### Phases
+
+We propose to divide the work in 3 phases. Each phase makes this work with
+either more isolation or more workloads. When no support is yet added to handle
+some workload, a clear error will be shown.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase.
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volume work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and works only when no volume or any
+combination of the following volume types are used by the pod:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 65535, mapping the range 0-65534 to the pod. This wide
```

Comment:
I'm curious why 64K?  The previous KEP had (I think) worked it out so you really didn't need that much.  But even if you did 64K, the old KEP still works and allows thousands of ranges at a time.

Reply:
The `nobody` user on most Linux systems is uid 65534.  Some systems use this user account directly, and in some circumstances it may be necessary to deal with unmapped uids (e.g. files on a volume owned by host uids that don't map to the container's user namespace), which are mapped to uid 65534 in the namespace.  So, size 65536 for a contiguous mapping that cover these use cases.

You can use multiple, non-contiguous mappings to dramatically reduce the number of host uids a workload needs, but it would be necessary to devise a way for the Pod spec to express which UIDs it needs mapped.  It would not be appropriate to make assumptions about what UIDs the workload needs.

Yes, 65536 is indeed enough for "most" workloads.  But if we enshrine that assumption, there will, inevitably, be some workloads that violate it.  If we're going to do this work, by all means use 65536 as a default but why figure out a way for the workload to express what it needs?  It will pay for itself down the track, when exotic workloads appear or when someone needs to run >65k Pods on a node.

Reply:
Is this something that we want to keep configurable so it allows modifying the range in the future? That opens the door to inspecting image and the runtime config to minimize the range or the workload expressing the size or range it needs.

Reply:
@thockin to make sure as much pods as possible are able to _enable_ userns. See the other responses on why IMHO that is important.

Also, this is what most of the tools give you out there. This is what you usually get with lxc/lxd, container runtimes, systemd-nspawn, etc., what posix kind of needs IIRC, for things to work fine. And things are working fine when using userns and 64k mappings.

@frasertweedale This is a very complex KEP already (see for example [this comment](https://github.com/kubernetes/enhancements/pull/2101#issuecomment-957696725) where I explain why it is quite complex). Adding more complexity now, for when "exotic workloads appear or when someone needs more than >65k pods in a node" is not a good reason to add the complexity _now_. We can add it later, when the need arise.

Reply:
@mrunalp IMHO the KEP is quite complicated today, I wouldn't add the complexity to express the UIDs the container needs in a reasonable API. I think that might be desirable in the future, but this is useful for quite a large use base as it is now.

What do you think? Also, we can always add it later, of course.

Reply:
@rata I agree it makes sense to defer such features if it helps get the initial support accepted sooner.  But, let's develop a schema that admits enhancements where, for example, the Pod spec expresses additional information about its requirements.  In other words, let's not use an Enum.  Instead, perhaps something along the lines of how volumes are specified:

Use host user namespace (i.e. today's behaviour and the **default** under this KEP):

```yaml
userNamespaceSpec:
  host: {}
```

Use the shared "static" allocation of size 65536, i.e. phase 1 and 2.  Maybe "shared" is a better identifier, or "cluster" (similar to previous version of this KEP).  As discussed, such a mode might eventually be deprecated.

```yaml
userNamespaceSpec:
  static: {}
```

Use a user namespace that maps host UIDs from a range assigned to the namespace in which the Pod lives (one of the suggestions for phase 3 of this KEP).  It is up to Kubelet (or the runtime) to decide HOW to map those IDs in the user namespace.

```yaml
userNamespaceSpec:
  namespace: {}
```

Similar to the previous example, but now the Pod spec expresses which UIDs it needs mapped.
The `requiredMappings` field reflects how [mappings are expressed in the OCI runtime spec)(https://github.com/opencontainers/runtime-spec/blob/main/config-linux.md#user-namespace-mappings), leaving Kubernetes to work out the host IDs.

```yaml
userNamespaceSpec:
  namespace:
    requiredMappings:
      - containerID: 0
        size: 1023
      - containerID: 65534
        size: 1
```

The advantages of this approach (or something like it):

- A single "top-level" field for the whole feature, instead of the two fields as currently proposed.
- No new "top-level" fields need to be introduced to enhance the capabilities in future phases.  You either introduce a new mode, or new fields for an existing mode.
- New modes, or enhancements to an existing mode, can be proposed, implemented and graduated separately.

Reply:
@frasertweedale thanks! I like the idea, but I'm not really sold on it. What do you think if we start with a bool to enable/disable userns (as ALL other namespaces are handled) and phase 1 and then, while that is on the works, we decide if we adopt something as you propose here?

We can later decide if we want something like this only, the bool and this (when the bool is enabled, some default mode can be used) or whatever.

But given that is not 100% clear how the next phases will look like, starting with something simple that has chances to stay (a bool, that is how we handle all namespaces) seems reasonable IMHO. We can later change it to something more complex, if all agree in the interface. We can even change it before it is included in a k8s release, we just need the basic understanding of how phase 1 should work now, then if it is one field that activates it or a section, it is not a big deal to change it.

Reply:
Added those things to the GA. So far I have:

 - Make plans on whether, when, and how to enable by default
 - Should we reconsider making the mappings smaller by default?
 - Should we allow any way for users to for "more" IDs mapped? If yes, how many more and how?
 - Should we allow the user can ask for specific mappings?

@thockin regarding not promising to allocate 64k, I'm not sure that will work. We can say "a number of user IDs determined by the kubelet", we can document it and all. And no matter what mapping range we use by default, it will have some length. Whenever we change the length, that is a user visible change, game over.

If apps worked before and not after we change the mapping length, with no visible changes in the pod.spec, it doesn't really matter if documentation says we never promised that. We will need to restore the previous behavior. And when we restore it, we will add a new value or field to enable this change. So, why not just do that in the first place?

We can try to play games like increasing the length is unlikely to cause issues, but decreasing the length is more likely to cause issues, but we are playing with fire. Likelihood is just a chance, not a guarantee. I think it makes sense in order to change the mapping range (a user visible change), we need to change the pod spec (add a new value to a field, a new field, etc.). And if you don't change the pod spec, then the user-visible stuff won't change.

You may want to make a change to take advantage of better isolation, etc. We might even do that by default after the field is present for several releases for example (so tools can start to set it to false not worrying about which k8s version a pod.spec will be used on). But I don't think trying to make user visible changes (i.e. breaking backwards compatibility) should be done unless we really have no other option, and we clearly have.

Reply:
Here's my thinking, but it doesn't need to block alpha.

If we hit beta at 64K, we can't really ever get smaller (without YET MORE controls in the API).  But if we start smaller, we can increase the default to cover a larger set of use-cases by default.  The point of beta is to get feedback, and we know we can't (initially) enable this by default.

So maybe "Make plans on whether, when, and how to enable by default" and "Should we reconsider making the mappings smaller by default?" should be beta criteria, rather than GA.

Reply:
@thockin that is exactly what I said, right? We can try to play that game, sure. I don't think that increasing the mapping length is significantly better than decreasing it.

If you increase/decrease the mapping length and go to a mapping length of X, it doesn't really mater if you are increasing or decreasing: the only thing that matters is if users that you care about work with length X. In both cases (increasing/decreasing) if the users that you are about are safe with X, all is well.

It is slightly better to increase for the reasons we both said already. I'm just not completely sold that is worth doing, though.

I'll add it to the unresolved things to discuss :)